orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java

```diff
@@ -0,0 +1,488 @@
+              .orElse(nestedField.name());
+          TypeDescription childType = buildOrcProjection(nestedField.fieldId(), nestedField.type(),
+              nestedField.isRequired(), mapping);
+          orcType.addField(name, childType);
+        }
+        break;
+      case LIST:
+        Types.ListType list = (Types.ListType) type;
+        TypeDescription elementType = buildOrcProjection(list.elementId(), list.elementType(),
+            list.isElementRequired(), mapping);
+        orcType = TypeDescription.createList(elementType);
+        break;
+      case MAP:
+        Types.MapType map = (Types.MapType) type;
+        TypeDescription keyType = buildOrcProjection(map.keyId(), map.keyType(), true, mapping);
+        TypeDescription valueType = buildOrcProjection(map.valueId(), map.valueType(), map.isValueRequired(),
+            mapping);
+        orcType = TypeDescription.createMap(keyType, valueType);
+        break;
+      default:
+        if (mapping.containsKey(fieldId)) {
+          TypeDescription originalType = mapping.get(fieldId).type();
+          Optional<TypeDescription> promotedType = getPromotedType(type, originalType);
+
+          if (promotedType.isPresent()) {
+            orcType = promotedType.get();
+          } else {
+            Preconditions.checkArgument(isSameType(originalType, type),
+                "Can not promote %s type to %s",
+                originalType.getCategory(), type.typeId().name());
+            orcType = originalType.clone();
+          }
+        } else {
+          if (isRequired) {
+            throw new IllegalArgumentException(
+                String.format("Field %d of type %s is required and was not found.", fieldId, type.toString()));
+          }
+
+          orcType = convert(fieldId, type, isRequired);
+        }
+    }
+
+    return orcType;
+  }
+
+  private static Map<Integer, OrcField> icebergToOrcMapping(String name, TypeDescription orcType,
+                                                            TypeUtil.NextID nextID) {
+    Map<Integer, OrcField> icebergToOrc = Maps.newHashMap();
+    switch (orcType.getCategory()) {
+      case STRUCT:
+        List<String> childrenNames = orcType.getFieldNames();
+        List<TypeDescription> children = orcType.getChildren();
+        for (int i = 0; i < children.size(); i++) {
+          icebergToOrc.putAll(icebergToOrcMapping(childrenNames.get(i), children.get(i), nextID));
+        }
+        break;
+      case LIST:
+        icebergToOrc.putAll(icebergToOrcMapping("element", orcType.getChildren().get(0), nextID));
+        break;
+      case MAP:
+        icebergToOrc.putAll(icebergToOrcMapping("key", orcType.getChildren().get(0), nextID));
+        icebergToOrc.putAll(icebergToOrcMapping("value", orcType.getChildren().get(1), nextID));
+        break;
+    }
+
+    if (orcType.getId() > 0) {
+      // Only add to non-root types.
+      int icebergId = icebergID(orcType).orElseGet(nextID::get);
```

Comment:
If the field id is missing, we seem to map based on the pre-order ID assignment. However these ids are completely different from Iceberg ids as they are not necessarily pre-order. This causes issues while reading the dataset as fields are not mapped correctly. IIRC in Avro at least, if the field id is missing for a particular field, we do not project the field. Should the same be done here? 
(In Avro, if file has no ids at all, then we do name-based matching or we also have to option to use name mapping, but that can be done separately).

Reply:
No, `nextID` is not based on pre-order (ORC),  it is computed from the max Iceberg ID used, unrelated to any ORC Id:
```java
AtomicInteger lastColumnId = new AtomicInteger(getMaxIcebergId(originalOrcSchema));
```
Where `getMaxIcebergId` traverses the iceberg ids stored in the file ORC schema.
There is precisely a test case covering this case, since the ID is from the max Iceberg ID used this does not cause mapping issues.

Reply:
I made an of error in my initial comment. When I said `pre-order`, I meant the traversal order of `icebergToOrcMapping` which I now realize is post-order. But I think the argument stands.

Consider:
Orc File Schema: 
1:a int 2:b int N/A: c struct< N/A: d int, N/A: e int>
where `N/A` represents that field id is missing.
Iceberg Schema:
1:a int 2: b int 4: c struct< 5: d int, 6: e int> 

In `buildOrcProjection`, lastColumnId = 2, thus the icebergToOrc mapping will now look like 
1:a int 2:b int 5: c struct< 3: d int, 4: e int> 
because of the post-order traversal in `icebergToOrcMapping`. Hence we see that the ID mappings are different between ORC and Iceberg schema. When we use this mapping to build the ORC projection, we will map id 4 in Iceberg schema which is a struct to `e` in the mapping which is an integer, which is incorrect.

I don't think we have the information to generate the correct mappings for fields missing IDs in the ORC file at this layer. I think we should probably handle this at a higher layer like we do with name mapping for Avro and leave fields with missing IDs as they are. https://github.com/apache/incubator-iceberg/blob/master/core/src/main/java/org/apache/iceberg/avro/ProjectionDatumReader.java#L55

Reply:
Yes, in this case we don't have enough information at this level hence we don't try to do name-based mapping here. I'm going to skip adding the column to the `icebergToOrc` mapping if no iceberg ID is found to avoid projection on that field. We can probably address fields with missing IDs in the ORC file with an approach like you mentioned, given possibly the option for name-based mapping as well for fallback.

Reply:
Thanks for making the changes!